File: 2024/day19.go
package main

import (
	"strings"

	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 19)

func possible(design string, patterns []string) bool {
	return utils.Any(len(patterns), func(i int) bool {
		rem := strings.TrimPrefix(design, patterns[i])
		return len(rem) < len(design) && (len(rem) == 0 || possible(rem, patterns))
	})
}

var memo = make(map[string]int)

func possibilities(design string, patterns []string) int {
	if _, ok := memo[design]; !ok {
		memo[design] = utils.Sum(len(patterns), func(i int) int {
			if p := patterns[i]; !strings.HasPrefix(design, p) {
				return 0
			} else if len(design) == len(p) {
				return 1
			} else {
				return possibilities(design[len(p):], patterns)
			}
		})
	}
	return memo[design]
}

func main() {
	groups := utils.Split(input, "\n\n")
	patterns := utils.Split(groups[0], ", ")
	designs := utils.Lines(groups[1])
	utils.Println(utils.Count(len(designs), func(i int) bool {
		return possible(designs[i], patterns)
	}))

	utils.Println(utils.Sum(len(designs), func(i int) int {
		return possibilities(designs[i], patterns)
	}))
}

File: 2024/rust/src/day19.rs
use std::collections::HashMap;

fn possible(design: &str, patterns: &[&str]) -> bool {
    patterns.iter().any(|&p| {
        let rem = design.strip_prefix(p).unwrap_or(design);
        rem.len() < design.len() && (design == p || possible(rem, patterns))
    })
}

fn possibilities(design: &str, patterns: &[&str], memo: &mut HashMap<String, usize>) -> usize {
    if let Some(&count) = memo.get(design) {
        count
    } else {
        let count = patterns
            .iter()
            .map(|p| match design.strip_prefix(p) {
                None => 0,
                Some("") => 1,
                Some(remaining) => possibilities(remaining, patterns, memo),
            })
            .sum();
        memo.insert(design.to_string(), count);
        count
    }
}

fn parse(input: &str) -> (Vec<&str>, Vec<&str>) {
    let groups = input.split("\n\n").collect::<Vec<&str>>();
    (groups[0].split(", ").collect(), groups[1].lines().collect())
}

pub fn part1(input: &str) -> String {
    let (patterns, designs) = parse(input);
    designs
        .iter()
        .filter(|d| possible(d, &patterns))
        .count()
        .to_string()
}

pub fn part2(input: &str) -> String {
    let (patterns, designs) = parse(input);
    designs
        .iter()
        .map(|d| possibilities(d, &patterns, &mut HashMap::new()))
        .sum::<usize>()
        .to_string()
}
